<a href="https://colab.research.google.com/github/Arifmulyaputra/IPSD-Assignment/blob/main/MODUL%205/Tugas_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

data_training = pd.read_csv('Training.csv')


data_head = data_training.head()
data_info = data_training.info()
data_description = data_training.describe(include='all')

data_head, data_info, data_description


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               2460 non-null   int64  
 1   Glucose                   2460 non-null   int64  
 2   BloodPressure             2460 non-null   int64  
 3   SkinThickness             2460 non-null   int64  
 4   Insulin                   2460 non-null   int64  
 5   BMI                       2460 non-null   float64
 6   DiabetesPedigreeFunction  2460 non-null   float64
 7   Age                       2460 non-null   int64  
 8   Outcome                   2460 non-null   int64  
dtypes: float64(2), int64(7)
memory usage: 173.1 KB


(   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
 0            6      148             72             35        0  33.6   
 1            1       85             66             29        0  26.6   
 2            8      183             64              0        0  23.3   
 3            1       89             66             23       94  28.1   
 4            0      137             40             35      168  43.1   
 
    DiabetesPedigreeFunction  Age  Outcome  
 0                     0.627   50        1  
 1                     0.351   31        0  
 2                     0.672   32        1  
 3                     0.167   21        0  
 4                     2.288   33        1  ,
 None,
        Pregnancies      Glucose  BloodPressure  SkinThickness      Insulin  \
 count  2460.000000  2460.000000    2460.000000    2460.000000  2460.000000   
 mean      3.817480   121.602033      68.915041      20.531301    80.119919   
 std       3.296458    31.789270      19.0826

Interpretasi Hasil EDA dan Pre-Processing
1. Struktur Data:
- Dataset memiliki 2.460 entri dengan 9 kolom, termasuk variabel target Outcome yang menunjukkan status diabetes (0 = tidak, 1 = ya).
- Semua fitur adalah numerik, sehingga tidak perlu encoding tambahan untuk variabel kategori.

2. Temuan EDA Awal:
- Nilai Tidak Realistis:
Kolom seperti Glucose, BloodPressure, SkinThickness, Insulin, dan BMI memiliki nilai 0, yang tidak mungkin dalam konteks medis.

- Distribusi Data:
  - Pregnancies: Rata-rata sekitar 3,8, dengan maksimum 17, menunjukkan data mencakup berbagai rentang usia wanita yang mungkin mengalami kehamilan.
  - Glucose: Rata-rata 122, dengan nilai minimum 0 yang diganti.
  - BloodPressure: Rata-rata 69, nilai minimum awal 0, yang tidak mungkin untuk tekanan darah manusia.
  - BMI: Memiliki rentang realistis setelah pre-processing, dengan nilai rata-rata 32,4.
  - Age: Usia berkisar antara 21 hingga 81 tahun, dengan median 29, yang menunjukkan dominasi populasi usia muda hingga paruh baya.

3. Pre-Processing yang Dilakukan:
- Nilai Hilang:
  - Semua nilai 0 pada kolom yang tidak realistis telah diganti dengan nilai median untuk menjaga distribusi data tetap seimbang.
  - Median dipilih untuk mengurangi dampak outlier seperti nilai insulin yang sangat tinggi (maksimum 846).

4. Implikasi untuk Analisis Lebih Lanjut:
- Modeling: Data siap untuk pelatihan model prediksi diabetes.
- Fitur Penting: Glucose, Insulin, BMI, dan Age tampak signifikan secara medis untuk klasifikasi diabetes.
- Normalisasi & Scaling: Disarankan untuk menormalkan data sebelum model tertentu seperti KNN atau SVM.

In [ ]:
# Import libraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Import algoritma klasifikasi
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

# Memisahkan fitur (X) dan target (y)
X = data_training.drop('Outcome', axis=1)
y = data_training['Outcome']

# Membagi data menjadi data latih dan uji (80% latih, 20% uji)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalisasi data untuk KNN
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Melatih dan mengevaluasi model
# KNN
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled, y_train)
knn_pred = knn.predict(X_test_scaled)
knn_acc = accuracy_score(y_test, knn_pred)

# Decision Tree
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)
dt_pred = dt.predict(X_test)
dt_acc = accuracy_score(y_test, dt_pred)

# Naive Bayes
nb = GaussianNB()
nb.fit(X_train, y_train)
nb_pred = nb.predict(X_test)
nb_acc = accuracy_score(y_test, nb_pred)

# Menampilkan hasil akurasi
print(f"Akurasi KNN: {knn_acc * 100:.2f}%")
print(f"Akurasi Decision Tree: {dt_acc * 100:.2f}%")
print(f"Akurasi Naive Bayes: {nb_acc * 100:.2f}%")


Akurasi KNN: 93.50%
Akurasi Decision Tree: 99.80%
Akurasi Naive Bayes: 71.75%


In [ ]:
# Import libraries tambahan untuk evaluasi
from sklearn.metrics import accuracy_score, recall_score, precision_score

# Fungsi evaluasi model
def evaluate_model(name, y_true, y_pred):
    acc = accuracy_score(y_true, y_pred)
    recall = recall_score(y_true, y_pred)
    precision = precision_score(y_true, y_pred)
    print(f"{name} Evaluation:")
    print(f" - Accuracy : {acc * 100:.2f}%")
    print(f" - Recall   : {recall * 100:.2f}%")
    print(f" - Precision: {precision * 100:.2f}%\n")
    return acc, recall, precision

# Evaluasi KNN
knn_acc, knn_recall, knn_precision = evaluate_model("KNN", y_test, knn_pred)

# Evaluasi Decision Tree
dt_acc, dt_recall, dt_precision = evaluate_model("Decision Tree", y_test, dt_pred)

# Evaluasi Naive Bayes
nb_acc, nb_recall, nb_precision = evaluate_model("Naive Bayes", y_test, nb_pred)


KNN Evaluation:
 - Accuracy : 93.50%
 - Recall   : 91.44%
 - Precision: 91.44%

Decision Tree Evaluation:
 - Accuracy : 99.80%
 - Recall   : 99.47%
 - Precision: 100.00%

Naive Bayes Evaluation:
 - Accuracy : 71.75%
 - Recall   : 57.75%
 - Precision: 64.29%



**Interpretasi dan Kesimpulan**

1. Accuracy: Mengukur seberapa sering model memprediksi dengan benar secara keseluruhan.

2. Recall (Sensitivity): Mengukur seberapa baik model mendeteksi kelas positif (penderita diabetes).

3. Precision: Mengukur seberapa banyak prediksi positif yang benar.

Setelah menjalankan kode ini, bandingkan ketiga metrik tersebut untuk setiap model:

- Model terbaik: Dipilih berdasarkan keseimbangan metrik accuracy, recall, dan precision, tergantung pada tujuan analisis (misalnya, jika mendeteksi diabetes, recall lebih penting untuk mengurangi false negative).